File: harness/src/sandbox/package-identity.ts
/**
 * The identity of a package, the query that asks for one, and the resolution
 * between the two.
 *
 * A package name crosses the harness, the embedder, and the provisioner. As a
 * bare `string` each site applies the rule that it believes, and two sites then
 * disagree on whether `decoupleR` and `decoupler` are one package. This module
 * is the one place that knows the rule.
 *
 * Two types carry the two ideas. A {@link PackageQuery} is what a person or an
 * agent asked: a spelling, an optional track, an optional version. A
 * {@link PackageIdentity} is the name that an ecosystem recognizes. The
 * identity is opaque, thus a caller cannot invent one, and every identity comes
 * from {@link pythonIdentity}, {@link rIdentity}, or {@link identityOf}.
 *
 * The PEP 503 fold lives in exactly one function of this file. The provisioner
 * holds the Python twin, `images/sandbox-provisioner/package_identity.py`, and
 * one fixture binds the two.
 */

import { err, ok, type Result } from "neverthrow";

/** The two ecosystems that the pool holds. */
export type Track = "python" | "r";

/**
 * What a person or an agent asked for. `spelling` is the name verbatim, as the
 * caller wrote it — a remedy quotes it, and an R name is case-sensitive at
 * `library()`. `track` names one ecosystem, and its absence searches both.
 * `version` pins one exact version.
 *
 * A query is never a key. Two queries of one fold are two asks, because a
 * caller that wrote `decoupleR` did not ask for `decoupler`.
 */
export type PackageQuery = {
    readonly spelling: string;
    readonly track?: Track;
    readonly version?: string;
};

/**
 * The brand of an identity name. It has no runtime value: it exists so that a
 * literal `{ track, name }` fails the typecheck outside this module, and thus
 * every identity comes from a constructor that applied the rule of its track.
 */
declare const identityBrand: unique symbol;

/**
 * The name that one ecosystem recognizes, under the track that holds it.
 *
 * The name of a Python identity is the PEP 503 fold of the spelling, because
 * PEP 503 defines the equivalence of a distribution name. The name of an R
 * identity is the DESCRIPTION spelling, verbatim, because `library()` is
 * case-sensitive. Two identities are equal when their keys are equal.
 */
export type PackageIdentity = {
    readonly track: Track;
    readonly name: string & { readonly [identityBrand]: true };
};

/**
 * The PEP 503 fold: each run of `-`, `_`, and `.` becomes one `-`, and the
 * case lowers. This is the ONE TypeScript home of the rule. The fold is
 * idempotent, thus a fold of a folded name gives that name again.
 */
function fold(name: string): string {
    return name.replace(/[-_.]+/g, "-").toLowerCase();
}

/**
 * Brand a name that a constructor of this module already ruled on.
 *
 * The cast is the whole point of the brand: the brand has no runtime value,
 * thus only a cast can produce one, and this function is the one place that
 * casts. A caller outside the module reaches a name through a constructor.
 */
function brand(name: string): PackageIdentity["name"] {
    // The brand is a compile-time marker with no runtime value, thus a cast is
    // the only way to mint one, and this function is the only site that casts.
    return name as PackageIdentity["name"];
}

/**
 * The Python identity of a spelling. The name folds, because PEP 503 makes
 * `PyYAML`, `pyyaml`, and `py_yaml` one distribution.
 *
 * @param spelling The distribution name, as the caller spells it.
 */
export function pythonIdentity(spelling: string): PackageIdentity {
    return { track: "python", name: brand(fold(spelling)) };
}

/**
 * The R identity of a name. The name keeps its spelling, because `library()`
 * is case-sensitive and `GO.db` loads under that exact name.
 *
 * @param name The DESCRIPTION `Package` value.
 */
export function rIdentity(name: string): PackageIdentity {
    return { track: "r", name: brand(name) };
}

/**
 * The identity of a name that an emitter already minted — a graph node, a lock
 * row. The dispatch is safe over such a name, because the fold is idempotent.
 *
 * @param track The ecosystem that holds the name.
 * @param name The name as the emitter wrote it.
 */
export function identityOf(track: Track, name: string): PackageIdentity {
    return track === "python" ? pythonIdentity(name) : rIdentity(name);
}

/**
 * The key of an identity: `<track>:<name>`. Two identities are equal when
 * their keys are equal, thus a map of identities keys on this string.
 */
export function identityKey(identity: PackageIdentity): string {
    return `${identity.track}:${identity.name}`;
}

/**
 * The identity that a key names, or `undefined` when the string is not a key.
 *
 * The FIRST colon splits the two halves. A track name holds no colon, and
 * neither a PEP 503 name nor an R name can hold one, thus `r:GO.db` gives the
 * R identity `GO.db`, dot and all. The name rides through {@link identityOf},
 * and the fold is idempotent, thus a key that {@link identityKey} wrote comes
 * back as the identity that wrote it.
 *
 * The key format is minted here, thus it is read here: a reader that splits
 * the string itself owns a second copy of the format.
 *
 * @param key A `<track>:<name>` string, as {@link identityKey} writes it.
 */
export function parseIdentityKey(key: string): PackageIdentity | undefined {
    const at = key.indexOf(":");
    if (at < 0) return undefined;
    const track = key.slice(0, at);
    const name = key.slice(at + 1);
    if (name === "") return undefined;
    if (track !== "python" && track !== "r") return undefined;
    return identityOf(track, name);
}

/**
 * The store address of an identity: the PEP 503 fold of its name, for both
 * tracks. A store directory is an address and not an identity, thus two
 * identities can share one address — `r:decoupleR` and `python:decoupler`
 * both address as `decoupler`. The pin marker inside the directory carries
 * the identity.
 */
export function identityAddress(identity: PackageIdentity): string {
    return fold(identity.name);
}

/**
 * Why an entry is not a query.
 *
 * - `empty` — the entry names no package.
 * - `location` — a path, a URL, or a store directory. A location is an
 *   installer detail, and the pool resolves names.
 * - `unknown_prefix` — a `<word>:` prefix that is neither `python:` nor `r:`.
 *   Unrefused, such a prefix rides into the pool as part of the name.
 * - `unsupported_specifier` — a specifier that is not `==`. Unrefused, a range
 *   such as `numpy>=1.26` becomes a package name.
 */
export type ParseQueryError =
    | { readonly type: "empty" }
    | { readonly type: "location"; readonly entry: string }
    | { readonly type: "unknown_prefix"; readonly prefix: string; readonly entry: string }
    | { readonly type: "unsupported_specifier"; readonly specifier: string; readonly entry: string };

/**
 * A path, a URL, or a store directory. The entry is trimmed already, thus a
 * leading space cannot defeat the guard.
 */
const LOCATION = /^[.~]|[/\\]|:\/\//;

/** The `<word>:` prefix of an entry. The group holds the word before the colon. */
const PREFIX = /^([A-Za-z][A-Za-z0-9_.-]*):/;

/** A run of specifier characters. Only the exact run `==` names a version. */
const SPECIFIER = /[<>!~=]+/;

/**
 * A run of characters that an exact version cannot hold: a specifier
 * character, a comma, or whitespace. `SPECIFIER` reads the FIRST run only,
 * thus `numpy==1.26,<2` gives the version `1.26,<2` without this guard, and
 * the compound range then rides into the pool as a version string.
 */
const VERSION_INTRUDER = /[<>!~=,\s]+/;

/**
 * Read one entry of the grammar `[python:|r:]<spelling>[==<version>]`.
 *
 * The entry trims once, and the parts keep what the trim left. Every reader of
 * the grammar in the harness calls this function: a second parser would
 * disagree with this one, and the plan validation that passed would then meet
 * a link pass that refuses.
 *
 * @param entry One package entry, as a plan, a tool call, or a host wrote it.
 */
export function parseQuery(entry: string): Result<PackageQuery, ParseQueryError> {
    const trimmed = entry.trim();
    if (trimmed === "") return err({ type: "empty" });
    if (LOCATION.test(trimmed)) return err({ type: "location", entry: trimmed });

    const prefix = PREFIX.exec(trimmed);
    const word = prefix?.[1];
    if (word !== undefined && word !== "python" && word !== "r") {
        return err({ type: "unknown_prefix", prefix: word, entry: trimmed });
    }
    const track: Track | undefined = word === "python" ? "python" : word === "r" ? "r" : undefined;
    const rest = track === undefined ? trimmed : trimmed.slice(track.length + 1);
    if (rest === "") return err({ type: "empty" });

    const specifier = SPECIFIER.exec(rest);
    if (specifier === null) return ok({ spelling: rest, ...(track === undefined ? {} : { track }) });
    if (specifier[0] !== "==") {
        return err({ type: "unsupported_specifier", specifier: specifier[0], entry: trimmed });
    }
    // Both halves trim: `numpy == 1.26.4` is one ask, and an untrimmed half
    // makes the spelling `numpy ` and the version ` 1.26.4`, neither of which
    // any index holds.
    const spelling = rest.slice(0, specifier.index).trim();
    if (spelling === "") return err({ type: "empty" });
    // `name==` names no version, thus the query pins none and the pool answers
    // the newest. The round-trip law holds, because `formatQuery` writes the
    // specifier only beside a version.
    const version = rest.slice(specifier.index + 2).trim();
    const intruder = VERSION_INTRUDER.exec(version);
    if (intruder !== null) {
        return err({ type: "unsupported_specifier", specifier: intruder[0], entry: trimmed });
    }
    return ok({
        spelling,
        ...(track === undefined ? {} : { track }),
        ...(version === "" ? {} : { version }),
    });
}

/**
 * Write a query in the grammar. The prefix rides only beside a track, and
 * `==<version>` only beside a version. For every query,
 * `parseQuery(formatQuery(query))` gives that query again.
 */
export function formatQuery(query: PackageQuery): string {
    const prefix = query.track === undefined ? "" : `${query.track}:`;
    const version = query.version === undefined ? "" : `==${query.version}`;
    return `${prefix}${query.spelling}${version}`;
}

/**
 * What a pool answers about the identities that it holds. Two reads are
 * enough for the ladder, thus a caller builds this over a graph, a lock, or a
 * census without exposing its own shape.
 */
export type PoolIndex = {
    /** Whether the pool holds this exact identity. */
    has: (identity: PackageIdentity) => boolean;
    /**
     * Each R identity of the pool whose address is this fold. The suggestion
     * of an unknown query comes from here: a person who wrote `seurat` meant
     * the R package `Seurat`.
     */
    rIdentitiesFoldingTo: (fold: string) => readonly PackageIdentity[];
};

/**
 * What a query resolves to.
 *
 * - `resolved` — one identity, thus the caller reads the pool under it.
 * - `ambiguous` — one spelling that both tracks hold. The caller refuses, and
 *   the remedy is a prefixed query.
 * - `unknown` — the pool holds no identity of this query. `suggestion` names
 *   the R identity that the caller probably meant.
 */
export type QueryResolution =
    | { readonly kind: "resolved"; readonly identity: PackageIdentity }
    | { readonly kind: "ambiguous"; readonly python: PackageIdentity; readonly r: PackageIdentity }
    | { readonly kind: "unknown"; readonly suggestion?: PackageIdentity };

/**
 * The sole R identity of the pool that folds to this fold, or none. Two such
 * identities give no suggestion, because a guess between them is a coin flip.
 */
function soleRSuggestion(pool: PoolIndex, spellingFold: string): PackageIdentity | undefined {
    const candidates = pool.rIdentitiesFoldingTo(spellingFold);
    return candidates.length === 1 ? candidates[0] : undefined;
}

/**
 * Resolve one query against a pool.
 *
 * The version takes no part, because the caller holds the versions: the graph
 * orders them, and this function answers which package the query names.
 *
 * The ladder, in order:
 *
 * 1. The query names a track: the identity of that track, when the pool holds
 *    it. Otherwise unknown, with the R suggestion when the track is `r`.
 * 2. The pool holds the R identity of the spelling and the Python identity of
 *    its fold: the R identity when the spelling differs from its fold, because
 *    an uppercase letter or a dot is evidence of an R spelling. Otherwise
 *    ambiguous.
 * 3. The pool holds the R identity: the R identity.
 * 4. The pool holds the Python identity: the Python identity.
 * 5. Exactly one R identity of the pool folds to the fold of the spelling:
 *    unknown, with that identity as the suggestion.
 * 6. Otherwise: unknown.
 *
 * A silent Python-first pick is a fault: it sends `decoupleR` to the Python
 * `decoupler`, and the R package never links.
 */
export function resolveQuery(query: PackageQuery, pool: PoolIndex): QueryResolution {
    const spellingFold = fold(query.spelling);

    if (query.track !== undefined) {
        const identity = identityOf(query.track, query.spelling);
        if (pool.has(identity)) return { kind: "resolved", identity };
        if (query.track !== "r") return { kind: "unknown" };
        const suggestion = soleRSuggestion(pool, spellingFold);
        return suggestion === undefined ? { kind: "unknown" } : { kind: "unknown", suggestion };
    }

    const r = rIdentity(query.spelling);
    const python = pythonIdentity(query.spelling);
    const holdsR = pool.has(r);
    const holdsPython = pool.has(python);

    if (holdsR && holdsPython) {
        return query.spelling === spellingFold ? { kind: "ambiguous", python, r } : { kind: "resolved", identity: r };
    }
    if (holdsR) return { kind: "resolved", identity: r };
    if (holdsPython) return { kind: "resolved", identity: python };

    const suggestion = soleRSuggestion(pool, spellingFold);
    return suggestion === undefined ? { kind: "unknown" } : { kind: "unknown", suggestion };
}

File: images/sandbox-provisioner/package_identity.py
#!/usr/bin/env python3
"""The identity of a package, the query that asks for one, and the grammar
between the two.

This module is the Python twin of `harness/src/sandbox/package-identity.ts`.
One conformance fixture, `harness/src/sandbox/__fixtures__/package-identity.json`,
binds the two: a case that one twin answers differently fails its own suite.

A `PackageQuery` is what a person or an agent asked: a spelling, an optional
track, an optional version. A `PackageIdentity` is the name that an ecosystem
recognizes. The name of a Python identity is the PEP 503 fold of the spelling,
because PEP 503 defines the equivalence of a distribution name. The name of an
R identity is the DESCRIPTION spelling, verbatim, because `library()` is
case-sensitive.

The module imports neither provision.py nor emit_deps.py. Both of those import
this one, thus the fold lives here once and no import cycle forces a second
copy.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# The two ecosystems that the pool holds.
TRACKS = ("python", "r")

# A path, a URL, or a store directory. A location is an installer detail, and
# the pool resolves names.
_LOCATION = re.compile(r"^[.~]|[/\\]|://")

# The `<word>:` prefix of an entry. The group holds the word before the colon.
_PREFIX = re.compile(r"^([A-Za-z][A-Za-z0-9_.-]*):")

# A run of specifier characters. Only the exact run `==` names a version.
_SPECIFIER = re.compile(r"[<>!~=]+")

# A run of characters that an exact version cannot hold: a specifier
# character, a comma, or whitespace. `_SPECIFIER` reads the FIRST run only,
# thus `numpy==1.26,<2` gives the version `1.26,<2` without this guard, and the
# compound range then rides into the pool as a version string.
_VERSION_INTRUDER = re.compile(r"[<>!~=,\s]+")


@dataclass(frozen=True)
class PackageIdentity:
    """The name that one ecosystem recognizes, under the track that holds it.

    Make one with `python_identity`, `r_identity`, or `identity_of`. Each of
    those applies the rule of its track, thus a name that reaches this class
    obeyed a rule.
    """
    track: str
    name: str


@dataclass(frozen=True)
class PackageQuery:
    """What a person or an agent asked for.

    `spelling` is the name verbatim, because a remedy quotes it and an R name
    is case-sensitive. `track` names one ecosystem, and None searches both.
    `version` pins one exact version.
    """
    spelling: str
    track: str | None = None
    version: str | None = None


class QueryError(Exception):
    """An entry that is not a query.

    `type` is one of `empty`, `location`, `unknown_prefix`, and
    `unsupported_specifier`. `prefix` rides with an unknown prefix, and
    `specifier` rides with an unsupported specifier.
    """

    def __init__(self, type: str, entry: str, prefix: str | None = None,
                 specifier: str | None = None):
        super().__init__(f"{type}: {entry!r}")
        self.type = type
        self.entry = entry
        self.prefix = prefix
        self.specifier = specifier


def _fold(name: str) -> str:
    """The PEP 503 fold: each run of `-`, `_`, and `.` becomes one `-`, and the
    case lowers. This is the ONE Python home of the rule. The fold is
    idempotent, thus a fold of a folded name gives that name again."""
    return re.sub(r"[-_.]+", "-", name).lower()


def python_identity(spelling: str) -> PackageIdentity:
    """The Python identity of a spelling. The name folds, because PEP 503 makes
    `PyYAML`, `pyyaml`, and `py_yaml` one distribution."""
    return PackageIdentity("python", _fold(spelling))


def r_identity(name: str) -> PackageIdentity:
    """The R identity of a name. The name keeps its spelling, because
    `library()` is case-sensitive and `GO.db` loads under that exact name."""
    return PackageIdentity("r", name)


def identity_of(track: str, name: str) -> PackageIdentity:
    """The identity of a name that an emitter already minted. The dispatch is
    safe over such a name, because the fold is idempotent."""
    if track == "python":
        return python_identity(name)
    if track == "r":
        return r_identity(name)
    raise ValueError(f"unknown track {track!r}; the tracks are {TRACKS}")


def key(identity: PackageIdentity) -> str:
    """The key of an identity: `<track>:<name>`. Two identities are equal when
    their keys are equal."""
    return f"{identity.track}:{identity.name}"


def parse_identity_key(key_text: str) -> PackageIdentity | None:
    """The identity that a key names, or None when the string is not a key.

    The FIRST colon splits the two halves. A track name holds no colon, and
    neither a PEP 503 name nor an R name can hold one, thus `r:GO.db` gives the
    R identity `GO.db`, dot and all. The name rides through `identity_of`, and
    the fold is idempotent, thus a key that `key` wrote comes back as the
    identity that wrote it.
    """
    at = key_text.find(":")
    if at < 0:
        return None
    track = key_text[:at]
    name = key_text[at + 1:]
    if not name or track not in TRACKS:
        return None
    return identity_of(track, name)


def address(identity: PackageIdentity) -> str:
    """The store address of an identity: the PEP 503 fold of its name, for both
    tracks.

    A store directory is an address and not an identity, thus two identities
    can share one address: `r:decoupleR` and `python:decoupler` both address as
    `decoupler`. The pin marker inside the directory carries the identity.
    """
    return _fold(identity.name)


def parse_query(entry: str) -> PackageQuery:
    """Read one entry of the grammar `[python:|r:]<spelling>[==<version>]`.

    The entry trims once, and the parts keep what the trim left. An entry that
    is not a query raises `QueryError`.
    """
    trimmed = entry.strip()
    if not trimmed:
        raise QueryError("empty", trimmed)
    if _LOCATION.search(trimmed):
        raise QueryError("location", trimmed)

    prefix = _PREFIX.match(trimmed)
    word = prefix.group(1) if prefix else None
    if word is not None and word not in TRACKS:
        raise QueryError("unknown_prefix", trimmed, prefix=word)
    track = word
    rest = trimmed if track is None else trimmed[len(track) + 1:]
    if not rest:
        raise QueryError("empty", trimmed)

    specifier = _SPECIFIER.search(rest)
    if specifier is None:
        return PackageQuery(rest, track, None)
    if specifier.group(0) != "==":
        raise QueryError("unsupported_specifier", trimmed,
                         specifier=specifier.group(0))
    # Both halves strip: `numpy == 1.26.4` is one ask, and an unstripped half
    # makes the spelling `numpy ` and the version ` 1.26.4`, neither of which
    # any index holds.
    spelling = rest[:specifier.start()].strip()
    if not spelling:
        raise QueryError("empty", trimmed)
    # `name==` names no version, thus the query pins none and the pool answers
    # the newest. The round-trip law holds, because `format_query` writes the
    # specifier only beside a version.
    version = rest[specifier.start() + 2:].strip()
    intruder = _VERSION_INTRUDER.search(version)
    if intruder is not None:
        raise QueryError("unsupported_specifier", trimmed,
                         specifier=intruder.group(0))
    return PackageQuery(spelling, track, version or None)


def format_query(query: PackageQuery) -> str:
    """Write a query in the grammar. The prefix rides only beside a track, and
    `==<version>` only beside a version. For every query,
    `parse_query(format_query(query))` gives that query again."""
    prefix = "" if query.track is None else f"{query.track}:"
    version = "" if query.version is None else f"=={query.version}"
    return f"{prefix}{query.spelling}{version}"
